### PR TITLE
Debugging

### PR DESCRIPTION
**Pull request by nilekhc**
**What this PR does / why we need it**:
Steps to setup local debugging for driver

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/version/version.go [hunk 1]
```diff
@@ -23,8 +23,8 @@ var (
 	Vcs string
 	// BuildTime is the date for the binary build
 	BuildTime string
-	// BuildVersion is the secrets-store-csi-driver version
-	BuildVersion string
+	// BuildVersion is the secrets-store-csi-driver version. Will be overwritten from build.
+	BuildVersion string = "v0.0.0"
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply pkg/version/version.go:27]: Set to `local-dev`

File: .gitignore [hunk 2]
```diff
@@ -78,3 +78,4 @@ cscope.*
 
 demo/
 docs/book/book
+__debug_bin
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .gitignore:81]: `Dlv` creates `__debug_bin`. When debugging, entire codebase is mounted as volume inside pod and when debugger is started it creates this debug binary at the root.

Hence, added this to gitignore so it won't be checked in accidently.

File: .local/README.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
+# Overview
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/README.md:1]: Good point. Added a note.

File: .local/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+WORKDIR /secrets-store-csi-driver-codebase
+
+COPY go.mod go.mod
+RUN go mod download
+
+EXPOSE 30123
+
+# these dlv debug arguments replicate driver args from DaemonSet
+ENTRYPOINT ["/go/bin/dlv", "--listen=:30123", "--accept-multiclient", "--headless=true", "--api-version=2", "debug", "./cmd/secrets-store-csi-driver", "--", "-v", "5", "-endpoint", "unix:///csi/csi.sock", "-nodeid", "kind-control-plane", "-enable-secret-rotation", "false", "-rotation-poll-interval", "30s", "-metrics-addr", ":8080", "-provider-volume", "/etc/kubernetes/secrets-store-csi-providers", "-filtered-watch-secret", "false"]
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/Dockerfile]: done

File: .local/debug-driver.yaml [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  namespace: kube-system
+spec:
+  type: NodePort
+  selector:
+    app: debug-driver
+  ports:
+  - name: debug
+    port: 30123
+    targetPort: 30123
+    nodePort: 30123
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/debug-driver.yaml:157]: done

File: .local/kind-config.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+- role: control-plane
+  extraPortMappings:
+  - containerPort: 30123 # node port for the delve server
+    hostPort: 30123 # port on your host machine to communicate with the delve server
+    protocol: TCP
+  extraMounts:
+    - hostPath: # /path/to/your/driver/secrets-store-csi-driver/codebase/on/host
+      containerPath: /secrets-store-csi-driver-codebase # path to the project folder inside cluster
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/kind-config.yaml]: done

File: .local/persistent-volume.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  name: driver-pv
+spec:
+  storageClassName: hostpath
+  capacity:
+    storage: 256Mi
+  accessModes:
+  - ReadWriteOnce
+  hostPath:
+    path: /secrets-store-csi-driver-codebase
+
```
Comment (aramase) [inline .local/persistent-volume.yaml:13]: nit: remove extra lines

File: .local/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,20 @@
+FROM golang:alpine
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/Dockerfile]: done

File: .local/README.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+It's much easier to debug code with breakpoints while developing new features or making changes to existing codebase. With this in mind, following steps provides a way to setup csi secret store driver for local debugging.
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/README.md]: done

File: .local/README.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+> NOTE: Steps in this guide are not tested by CI/CD. This is just one of the way to locally debug the code and a good starting point.
+
+## Prerequisites
+
+* [Docker Desktop](https://docs.docker.com/get-docker)
+* [kind (Kubernetes in Docker)](https://kind.sigs.k8s.io)
+* [Kubectl](https://kubernetes.io/de/docs/tasks/tools/install-kubectl)
+* [Visual Studio Code](https://code.visualstudio.com/download)
+* [GOLang](https://golang.org/doc/install)
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/README.md]: done

File: .local/README.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+```
+
+### Deploy resources for debugging
+- Deploy following Driver resources:
+```sh
+kubectl apply -f deploy/rbac-secretproviderclass.yaml
+kubectl apply -f deploy/csidriver.yaml
+kubectl apply -f deploy/secrets-store.csi.x-k8s.io_secretproviderclasses.yaml
+kubectl apply -f deploy/secrets-store.csi.x-k8s.io_secretproviderclasspodstatuses.yaml
+kubectl apply -f deploy/rbac-secretprovidersyncing.yaml
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/README.md:48]: Good point. Added.

File: .local/README.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+# deploy your specific provider
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/README.md]: added

File: .local/debug-driver.yaml [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+          hostPath:
+            path: /var/lib/kubelet/plugins/csi-secrets-store/
+            type: DirectoryOrCreate
+        - name: providers-dir
+          hostPath:
+            path: /etc/kubernetes/secrets-store-csi-providers
+            type: DirectoryOrCreate
+
+---
+#Separate deployment to manage actual driver debugging independent of node registrar and livenessprobe
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/debug-driver.yaml]: done

File: .local/persistent-volume.yaml (hunk 7 above)
Comment (nilekhc) [reply .local/persistent-volume.yaml]: done

File: .local/debug-driver.yaml [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: debug-driver
+  namespace: kube-system
+spec:
+  replicas: 1 #Keep single replica for debugging
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .local/debug-driver.yaml]: done

**Review by aramase — COMMENTED** [review]
/lgtm

**Pull request merged**